src/mame/casio/cz1.cpp

```diff
@@ -0,0 +1,924 @@
+void cz1_state::main_to_sub_1_w(u8 data)
+{
+	machine().scheduler().synchronize();
+	m_main_to_sub[1] = data;
+}
+
+/**************************************************************************/
+u8 cz1_state::main_to_sub_1_r()
+{
+	return m_main_to_sub[1];
+}
+
+/**************************************************************************/
+void cz1_state::sub_to_main_w(u8 data)
+{
+	machine().scheduler().synchronize();
+	m_sub_to_main = data;
+}
+
+/**************************************************************************/
+u8 cz1_state::sub_to_main_r()
+{
+	return m_sub_to_main;
+}
+
+/**************************************************************************/
+int cz1_state::cont_r()
+{
+	return BIT(m_main_port[1], 3);
+}
+
+/**************************************************************************/
+int cz1_state::sync_r()
+{
+	return m_sync;
+}
+
+/**************************************************************************/
+void cz1_state::sync_clr_w(u8)
+{
+	machine().scheduler().synchronize();
+	m_sync = 0;
+}
+
+/**************************************************************************/
+void cz1_state::main_irq_w(u8)
+{
+	m_maincpu->set_input_line(UPD7810_INTF1, ASSERT_LINE);
+}
+
+/**************************************************************************/
+void cz1_state::main_irq_ack_w(u8)
+{
+	m_maincpu->set_input_line(UPD7810_INTF1, CLEAR_LINE);
+}
+
+/**************************************************************************/
+u8 cz1_state::mcu_r()
+{
+	if (!machine().side_effects_disabled())
+		m_sync49 = 0;
+
+	return ~m_mcu->p1_r();
+}
+
+/**************************************************************************/
+void cz1_state::mcu_p2_w(u8 data)
+{
+	if (BIT(data ^ m_mcu_p2, 6))
+		m_maincpu->set_input_line(UPD7810_INTF2, BIT(~data, 6));
+
+	if (BIT(~data & m_mcu_p2, 7))
+		m_sync49 = 1;
+
+	m_mcu_p2 = data;
+}
+
+/**************************************************************************/
+int cz1_state::cont49_r()
+{
+	return BIT(m_mcu_p2, 5);
+}
+
+/**************************************************************************/
+int cz1_state::sync49_r()
+{
+	return m_sync49;
+}
+
+/**************************************************************************/
+void cz1_state::machine_start()
+{
+	m_leds.resolve();
+	m_led_env.resolve();
+	m_led_bank.resolve();
+	m_led_tone.resolve();
+
+	// aftertouch amp levels (TODO: are these correct?)
+	for (int i = 0; i < 0x40; i++)
+		m_volume[i] = pow(2, (float)i / 0x3f) - 1.0;
+
+	// register for save states
+	save_item(NAME(m_main_port));
+	save_item(NAME(m_mcu_p2));
+	save_item(NAME(m_midi_rx));
+	save_item(NAME(m_cart_addr));
+	save_item(NAME(m_main_to_sub));
+	save_item(NAME(m_sub_to_main));
+	save_item(NAME(m_sync));
+	save_item(NAME(m_sync49));
+}
+
+/**************************************************************************/
+void cz1_state::machine_reset()
+{
+	m_main_port[0] = m_main_port[1] = m_main_port[2] = 0xff;
+	m_mcu_p2 = 0xff;
```

Comment:
You shouldn’t preemptively change the MCU output values on reset – doing that makes things get out of sync.  If the MCU updates its outputs on reset, it will call you to tell you.